File: 백준/Gold/5557. 1학년/1학년.js
const rl = require('readline').createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];

rl.on('line', (line) => {
  input.push(line);
}).on('close', () => {
  const answer = solution(input);
  console.log(answer);
  process.exit(0);
});

const solution = ([size, input]) => {
  const N = Number(size);
  let numbers = input.split(' ').map(Number);

  const dp = Array.from({ length: N - 1 }, () => new Array(21).fill(0n));
  dp[0][numbers[0]] = 1n;

  for (let i = 1; i < N - 1; ++i) {
    const number = numbers[i];
    const push = (n, count) => {
      if (n >= 0 && n <= 20) {
        dp[i][n] += count;
      }
    };

    dp[i - 1].forEach((count, sum) => {
      push(sum + number, count);
      push(sum - number, count);
    });
  }

  return "" + dp[N - 2][numbers[N - 1]];
};

File: 백준/Gold/5557. 1학년/1학년.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        String[] input = br.readLine().split(" ");
        int[] numbers = new int[N];

        for (int i = 0; i < N; i++) {
            numbers[i] = Integer.parseInt(input[i]);
        }

        System.out.println(solution(N, numbers));
    }

    public static long solution(int N, int[] numbers) {
        Map<Integer, Long>[] dp = new HashMap[N - 1];

        for (int i = 0; i < N - 1; i++) {
            dp[i] = new HashMap<>();
        }

        dp[0].put(numbers[0], 1L);

        for (int i = 1; i < N - 1; i++) {
            int number = numbers[i];
            for (Map.Entry<Integer, Long> entry : dp[i - 1].entrySet()) {
                int sum = entry.getKey();
                long count = entry.getValue();

                push(dp[i], sum + number, count);
                push(dp[i], sum - number, count);
            }
        }

        return dp[N - 2].getOrDefault(numbers[N - 1], 0L);
    }

    private static void push(Map<Integer, Long> map, int n, long count) {
        if (n >= 0 && n <= 20) {
            map.put(n, map.getOrDefault(n, 0L) + count);
        }
    }
}
